algorithms/C/linked-lists/glthreads.c

```diff
@@ -0,0 +1,133 @@
+/*
+In traditional a doubly linked-list each node will have a left and right
+pointer and a pointer to the data, so the structure of the node will contain
+3 data members(next, prev, data). In Glue based doubly linked-list there will
+only be 2 data members left and right pointers. As for the data, it will be 
+glued on top of the node.
+
+Illustration
+-------------
+Traditional doubly linked-list:
+                         _____________                    _____________
+                    ___ | application |               ___| application |
+                   |    |   data      |              |   |    data     |
+                   |    |_____________|              |   |_____________|
+         __________|_________              __________|_________
+        | prev | data | next |____________| prev | data | next |
+        |______|______|______|            |______|______|______|
+
+
+Glue based doubly linked-list:
+        
+         _____________             _____________
+        | application |           | application |
+        |    data     |           |    data     |
+        |_____________|           |_____________|
+        | prev | next |___________| prev | next |
+        |______|______|           |______|______|
+
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#define HEAD(list) (list->head)
+#define OFFSETOF(structure, field) (long)&((structure*)0)->field
+#define ITERATE_GL_THREADS_BEGIN(lstptr, struct_type, ptr)      \
+{                                                               \
+    glthread_node *_glnode = NULL, *_next = NULL;               \
+    for(_glnode = lstptr->head; _glnode; _glnode = _next){      \
+        _next = _glnode->next;                                  \
+        ptr = (struct_type *)((char *)_glnode - lstptr->offset);
+#define ITERATE_GL_THREADS_ENDS }}
+
+typedef struct glthread_node_{
+    struct glthread_node_* next;
+    struct glthread_node_* prev;
+}glthread_node;
+
+typedef struct{
+   glthread_node* head;
+   unsigned int offset;
+}gldll;
+
+void init_glthread_node(glthread_node* node){
+    node->next = NULL;
+    node->prev = NULL;
+}
+
+void init_glthread_list(gldll* list, unsigned int offset){
+    list->head = NULL;
+    list->offset = offset;
+}
+
+void glthread_pushfront(gldll* list, glthread_node* node){
+    glthread_node* cur = HEAD(list);
+    if(cur == NULL){
+        HEAD(list) = node;
+        return;
+    }
+    cur->prev = node;
+    node->next = cur;
+    HEAD(list) = node;
+}
+
+typedef struct{
+    char name[30];
+    int age, weight, height;
+    glthread_node glnode;
+}person;
+
+void print(person* per){
+    printf("Name: %s\n", per->name);
+    printf("Height: %i\n", per->height);
+    printf("Age: %i\n", per->age);
+    printf("Weight: %i\n", per->weight);
+	printf("------------\n");
+}
+
+int main(){
+    gldll* list = malloc(sizeof(*list));
+
+    person* rahul = malloc(sizeof(*rahul));
+    person* rohit = malloc(sizeof(*rohit));
```

Comment:
@aayushjain7 are you sure you used gcc or clang as the compiler and not g++?

Reply:
`gcc` and `g++` are same commands that just invoke the gcc driver, the actual compiler itself is chosen based on the type of file itself ( until the compiler isn't told to do so with `-x` option ).

Looking at the error message, it looks like on their local machine, @aayushjain7 accidentally changed the extension to `.cpp`, making it begin treated as C++ file and not a C file. 

C++ ( unlike C ), enforces much more type safety and doesn't allow program to implicitly convert a void pointer to something else hence the error.

Reply:
but both are a little different right without any flags?

I read that `g++` treats both .c and .cpp file as only C++ files whereas `gcc` considers .c and .cpp file differently as C and C++ files respectively. I was able to replicate the same errors when i use `g++` in a .c file.

I used to explicitly convert a `void*` to the required pointer type but somewhere in quora or stackoverflow i read that the type casting is bad at times, as it can give rise to new bugs and something about readability issues? and many good C devs avoid that or something. With that and knowing that the compiler does it for us so i stopped doing it in C.

Reply:
`g++` just implicitly links C++ standard library for you. If you indent to compile C++ program with `gcc`, you then have to perform that linking step manually. The actual compiler it

In C, it is recommended to let the conversion happen implicitly.

---

FWIW If you are really interested in how GCC compiles your program, you can always use a debugger on the compiler using the following command ( using `gdb` for demonstration purposes, you can also do the same with any other debugger )

```
$ gcc <parameters> -wrapper gdb,--args
```

Although for much detailed debugging session, you must have a gcc itself built with debugging info enabled ( that is a whole different story )